ksqldb-engine/src/main/java/io/confluent/ksql/engine/TransientQueryCleanupService.java

```diff
@@ -0,0 +1,276 @@
+import org.slf4j.LoggerFactory;
+
+public class TransientQueryCleanupService extends AbstractScheduledService {
+  private static final Logger LOG = LoggerFactory.getLogger(TransientQueryCleanupService.class);
+  private static final Pattern TRANSIENT_PATTERN =
+          Pattern.compile("(?i).*transient_.*_[0-9]\\d*_[0-9]\\d*");
+  private static final int NUM_OF_RETRIES_TO_DELETE_RESOURCE = 5;
+
+  private final BlockingQueue<Callable<Boolean>> cleanupTasks;
+  private final Retryer<Boolean> retryer;
+  private final String stateDir;
+  private final ServiceContext serviceContext;
+  private final Set<TransientQueryMetadata> allTransientQueriesEverCreated;
+  private final Set<String> allTransientQueriesEverCreatedIds;
+  private final int initialDelay;
+  private final int intervalPeriod;
+
+
+  public TransientQueryCleanupService(final ServiceContext serviceContext,
+                                      final KsqlConfig ksqlConfig) {
+    this.initialDelay = ksqlConfig.getInt(
+            KsqlConfig.KSQL_TRANSIENT_QUERY_CLEANUP_SERVICE_INITIAL_DELAY_SECONDS);
+
+    this.intervalPeriod = ksqlConfig.getInt(
+            KsqlConfig.KSQL_TRANSIENT_QUERY_CLEANUP_SERVICE_PERIOD_SECONDS);
+
+    cleanupTasks = new LinkedBlockingDeque<>();
+
+    retryer = RetryerBuilder.<Boolean>newBuilder()
+            .retryIfResult(aBoolean -> Objects.equals(aBoolean, false))
+            .retryIfExceptionOfType(IOException.class)
+            .retryIfRuntimeException()
+            .withStopStrategy(StopStrategies.stopAfterAttempt(NUM_OF_RETRIES_TO_DELETE_RESOURCE))
+            .build();
+
+    this.stateDir = ksqlConfig.getKsqlStreamConfigProps()
+            .getOrDefault(
+                    StreamsConfig.STATE_DIR_CONFIG,
+                    StreamsConfig.configDef()
+                            .defaultValues()
+                            .get(StreamsConfig.STATE_DIR_CONFIG))
+            .toString();
+    this.serviceContext = serviceContext;
+    this.allTransientQueriesEverCreated = new HashSet<>();
+    this.allTransientQueriesEverCreatedIds = new HashSet<>();
+  }
+
+  @Override
+  protected void runOneIteration() {
+    LOG.info("Starting cleanup for leaked resources.");
+
+    findCleanupTasks();
+
+    final long numLeakedTopics = cleanupTasks.stream()
+            .filter(p -> p.getClass().equals(TransientQueryTopicCleanupTask.class))
+            .count();
+
+    final long numLeakedStateDirs = cleanupTasks.stream()
+            .filter(p -> p.getClass().equals(TransientQueryStateCleanupTask.class))
+            .count();
+
+    LOG.info("Cleaning up {} leaked topics.", numLeakedTopics);
+    LOG.info("Cleaning up {} leaked state directories.", numLeakedStateDirs);
+    try {
+      while (!cleanupTasks.isEmpty()) {
+        final Callable<Boolean> task = cleanupTasks.take();
+
+        try {
+          retryer.call(task);
+        } catch (RetryException | ExecutionException e) {
+          cleanupTasks.add(task);
```

Comment:
If you fail for a given task, it doesn't seem that there's any backoff, so it'll just hammer Kafka 5 times in a row and then be done.  If you used a PriorityQueue and had time to run, you could peek at the top task and sleep until it was ready.  Then you could make each task adjustable and have X seconds between attempts.

Another simpler approach is that at each call to `runOneIteration`, you can go through everything once, but if you get to a task that failed in this round, you declare yourself done and wait until the next iteration.

Reply:
changed. Upon failure, it doesn't retry but instead, queues the task back again

Reply:
I think if a deleteion task fails at a moment, then it is less probably to fail after ~10 minutes as compared to immediately. Not the most scientific reasoning but a good enough heuristic